### PR TITLE
(WIP) Liqun/nuphar shape infer

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: orttraining/orttraining/core/session/training_session.cc [hunk 1]
```diff
@@ -215,6 +215,12 @@ Status TrainingSession::ConfigureForTraining(
 
   ORT_RETURN_IF_ERROR(ApplyTransformationsToMainGraph(trainable_initializers, config.enable_gelu_approximation));
 
+  // std::string filename = "/bert_ort/liqun/test_out/mc_post_apply_trx.onnx";
+  // Save(filename, SaveOption::NO_RELOAD);
+
+  common::Status status = Model::Load("/bert_ort/liqun/test_out/mc_post_apply_trx_post_shape_infer.onnx", model_,
```
Comment (ytaous) [inline orttraining/orttraining/core/session/training_session.cc:221]: >liqun [](start = 49, length = 5)

You'' remove/update this later, right?

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -16,6 +16,14 @@
 from .checkpointing_utils import list_checkpoint_files, get_checkpoint_name, CombineZeroCheckpoint
 import onnxruntime.capi.pt_patch
 
+import sys
+sys.path.append('/bert_ort/liqun/onnxruntime/onnxruntime/core/providers/nuphar/scripts')
```
Comment (ytaous) [inline orttraining/orttraining/python/ort_trainer.py:20]: >liqun [](start = 27, length = 5)

same comment

File: orttraining/orttraining/python/ort_trainer.py [hunk 3]
```diff
@@ -674,6 +682,13 @@ def _init_session(self):
             self._extra_postprocess(self.onnx_model_)
 
         self._verify_fully_optimized_model(self.onnx_model_)
+
+        pre_shape_infer_model_path = '/bert_ort/liqun/test_out/mc_pre_shape_infer_2.onnx'
```
Comment (ytaous) [inline orttraining/orttraining/python/ort_trainer.py:686]: >liqun [](start = 48, length = 5)

same

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml [hunk 4]
```diff
@@ -6,7 +6,7 @@ trigger: none
 jobs:
 - job: Onnxruntime_Linux_GPU_Training_FrontEnd
 
-  timeoutInMinutes: 120
+  timeoutInMinutes: 240
```
Comment (ytaous) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml:9]: >240 [](start = 20, length = 3)

you need 4 hours?

File: orttraining/orttraining/core/session/training_session.cc (hunk 1 above)
Comment (liqunfu) [reply orttraining/orttraining/core/session/training_session.cc:221]: Yes.
This is the main thing to be solved if we want to the nuphar inference tool for training.

File: orttraining/orttraining/python/ort_trainer.py (hunk 2 above)
Comment (liqunfu) [reply orttraining/orttraining/python/ort_trainer.py:20]: need to package the symbolic_shape_infer with training.

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml (hunk 4 above)
Comment (liqunfu) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-frontend-test-ci-pipeline.yml:9]: each multiple choice finetune takes more than an hour.

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 5]
```diff
@@ -1038,6 +1041,38 @@ def _infer_Slice(self, node):
             assert len(ends) == 1
             self.sympy_data_[node.output[0]] = self.sympy_data_[node.input[0]][starts[0]:ends[0]]
 
+    def _infer_Gelu(self, node):
+        data_shape = self._get_shape(node, 0)
+        vi = self.known_vi_[node.output[0]]
+        vi.CopyFrom(helper.make_tensor_value_info(node.output[0], onnx.TensorProto.FLOAT, data_shape))
+        self.known_vi_[vi.name] = vi
```
Comment (KeDengMS) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:1048]: >self.known_vi_[vi.name] = vi [](start = 8, length = 28)

seems not needed? vi.name is node.output[0], which matches when vi is assigned from self.known_vi_

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 6]
```diff
@@ -1038,6 +1041,38 @@ def _infer_Slice(self, node):
...
+
+        data_shape[len(data_shape) - 1] = 1
+        vi = self.known_vi_[node.output[1]]
+        vi.CopyFrom(helper.make_tensor_value_info(node.output[1], onnx.TensorProto.FLOAT, data_shape))
+        self.known_vi_[vi.name] = vi
+
+        vi = self.known_vi_[node.output[2]]
+        vi.CopyFrom(helper.make_tensor_value_info(node.output[2], onnx.TensorProto.FLOAT, data_shape))
+        self.known_vi_[vi.name] = vi
+
```
Comment (KeDengMS) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:1064]: From schema, seems output 1/2 is optional

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 7]
```diff
@@ -1038,6 +1041,38 @@ def _infer_Slice(self, node):
...
+    def _infer_SoftmaxCrossEntropyLoss(self, node):
+        vi = self.known_vi_[node.output[0]]
+        vi.type.tensor_type.elem_type = onnx.TensorProto.FLOAT
```
Comment (KeDengMS) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:1067]: >onnx.TensorProto.FLOAT [](start = 40, length = 22)

should this come from input elem_type?

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 8]
```diff
@@ -1038,6 +1041,38 @@ def _infer_Slice(self, node):
...
+        self.known_vi_[vi.name] = vi
+
+        if len(node.output) > 1:
+            data_shape = self._get_shape(node, 0)
+            vi = self.known_vi_[node.output[1]]
+            vi.CopyFrom(helper.make_tensor_value_info(vi.name, onnx.TensorProto.FLOAT, data_shape))
+            self.known_vi_[vi.name] = vi
```
Comment (KeDengMS) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:1074]: May add input/output index to self._pass_on_shape_and_type() and reuse that code.

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 9]
```diff
@@ -1038,6 +1041,38 @@ def _infer_Slice(self, node):
...
+    def _infer_Gelu(self, node):
+        data_shape = self._get_shape(node, 0)
+        vi = self.known_vi_[node.output[0]]
+        vi.CopyFrom(helper.make_tensor_value_info(node.output[0], onnx.TensorProto.FLOAT, data_shape))
+        self.known_vi_[vi.name] = vi
+
+    def _infer_LayerNormalization(self, node):
+        data_shape = self._get_shape(node, 0)
+        vi = self.known_vi_[node.output[0]]
+        vi.CopyFrom(helper.make_tensor_value_info(node.output[0], onnx.TensorProto.FLOAT, data_shape))
```
Comment (KeDengMS) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:1053]: >onnx.TensorProto.FLOAT [](start = 66, length = 22)

hardcoded type may have issue with float16/double?

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 10]
```diff
@@ -1153,6 +1188,12 @@ def _infer_impl(self, in_mp, start_sympy_data={}):
 
         for node in self.out_mp_.graph.node:
             assert all([i in self.known_vi_ for i in node.input if i])
+            # if node.op_type == 'SoftmaxCrossEntropyLoss':
+            #     # import pdb; pdb.set_trace()
+            #     print("")
+            if node.op_type == 'LayerNormalization':
+                # import pdb; pdb.set_trace()
+                print("")
```
Comment (KeDengMS) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:1196]: Please clean up debug code

**Pull request closed without merging**